testing/base/source_gen_test.cpp

```diff
@@ -0,0 +1,173 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "testing/base/source_gen.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "common/set.h"
+#include "testing/base/gtest_main.h"
+#include "toolchain/driver/driver.h"
+
+namespace Carbon::Testing {
+namespace {
+
+using ::testing::AllOf;
+using ::testing::ContainerEq;
+using ::testing::Contains;
+using ::testing::Eq;
+using ::testing::Ge;
+using ::testing::Le;
+using ::testing::MatchesRegex;
+using ::testing::SizeIs;
+
+template <typename T>
+static auto SumSizes(const T& range) -> ssize_t {
+  return std::accumulate(
+      range.begin(), range.end(), static_cast<ssize_t>(0),
+      [](ssize_t lhs, const auto& rhs) -> ssize_t { return lhs + rhs.size(); });
+}
+
+TEST(SourceGenTest, UniqueIds) {
+  SourceGen gen;
+
+  auto unique = gen.GetShuffledUniqueIds(1000);
+  EXPECT_THAT(unique.size(), Eq(1000));
+  Set<llvm::StringRef> set;
+  for (llvm::StringRef id : unique) {
+    EXPECT_THAT(id, MatchesRegex("[A-Za-z][A-Za-z0-9_]*"));
+    EXPECT_TRUE(set.Insert(id).is_inserted()) << "Colliding id: " << id;
+  }
+
+  // Check that repeated calls are different in interesting ways, but have the
+  // exact same total bytes.
+  ssize_t unique_size_sum = SumSizes(unique);
+  for ([[maybe_unused]] int i : llvm::seq(1, 10)) {
+    auto unique2 = gen.GetShuffledUniqueIds(1000);
+    EXPECT_THAT(unique2, SizeIs(1000));
+    // Should be (at least) a different shuffle of identifiers.
+    EXPECT_THAT(unique2, Not(ContainerEq(unique)));
+    // But the sum of lengths should be identical.
+    EXPECT_THAT(SumSizes(unique2), Eq(unique_size_sum));
+  }
+
+  // Check length constraints have the desired effect.
+  unique = gen.GetShuffledUniqueIds(1000, /*min_length=*/10, /*max_length=*/20);
+  for (llvm::StringRef id : unique) {
+    EXPECT_THAT(id.size(), Ge(10)) << "Too short id: " << id;
+    EXPECT_THAT(id.size(), Le(20)) << "Too long id: " << id;
+  }
```

Comment:
Had you considered:

```suggestion
  EXPECT_THAT(unique, Each(SizeIs(AllOf(Ge(10), Le(20)))));
```

Or if you prefer the loop for id printing:

```suggestion
  for (llvm::StringRef id : unique) {
    EXPECT_THAT(id, SizeIs(AllOf(Ge(10), Le(20))));
  }
```